lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java

```diff
@@ -0,0 +1,217 @@
+
+    connectionFactoryMap.put(connFactoryId, new NSConnectionFactory<>(this, connFactoryId, codec, eventHandler, linkListener));
+    isStreamingCodecMap.put(connFactoryId, codec instanceof StreamingCodec);
+  }
+
+  @Override
+  public synchronized void unregisterConnectionFactory(final Class<? extends Name<String>> connectionFactoryId) {
+    final String connFactoryId = connectionFactoryId.getName();
+
+    final ConnectionFactory  connFactory = connectionFactoryMap.get(connFactoryId);
+    if (connFactory != null) {
+      connectionFactoryMap.remove(connFactoryId);
+      isStreamingCodecMap.remove(connFactoryId);
+    }
+  }
+
+  /**
+   * Registers identifier of NetworkService.
+   * @param nsId
+   * @throws Exception
+   */
+  @Override
+  public void registerId(final Identifier nsId) {
+    LOG.log(Level.INFO, "Registering NetworkSerice " + nsId);
+    this.myId = nsId;
+    final Tuple<Identifier, InetSocketAddress> tuple =
+        new Tuple<>(nsId, (InetSocketAddress) this.transport.getLocalAddress());
+    LOG.log(Level.FINEST, "Binding {0} to NetworkService@({1})",
+        new Object[]{tuple.getKey(), tuple.getValue()});
+    this.nameServiceRegisteringStage.onNext(tuple);
+  }
+
+  /**
+   * Open a channel for remoteId
+   * @param remoteId
+   * @throws NetworkException
+   */
+  <T> Link<DefaultNSMessage<T>> openLink(final Identifier remoteId) throws NetworkException {
+    try {
+      final SocketAddress address = nameResolver.lookup(remoteId);
+      if (address == null) {
+        throw new NetworkException("Lookup " + remoteId + " is null");
+      }
+      return transport.open(address, nsCodec, nsLinkListener);
+    } catch(Exception e) {
+      e.printStackTrace();
+      throw new NetworkException(e);
+    }
+  }
+
+  /**
+   * Gets a ConnectionFactory
+   * @param connectionFactoryId the identifier of the ConnectionFActory
+   */
+
+  @Override
+  public <T> ConnectionFactory<T> getConnectionFactory(final Class<? extends Name<String>> connectionFactoryId) {
+
+    final ConnectionFactory<T> connFactory = connectionFactoryMap.get(connectionFactoryId.getName());
+
+    if (connFactory == null) {
+      throw new RuntimeException("Cannot find ConnectionFactory of " + connectionFactoryId + ".");
+    }
+
+    return connFactory;
+  }
+
+  @Override
+  public void unregisterId(final Identifier nsId) {
+    this.myId = null;
+    LOG.log(Level.FINEST, "Unbinding {0} to NetworkService@({1})",
+        new Object[]{nsId, this.transport.getLocalAddress()});
+    this.nameServiceUnregisteringStage.onNext(nsId);
+  }
+
+  @Override
+  public Identifier getNetworkServiceId() {
+    return this.myId;
+  }
+
+  @Override
+  public SocketAddress getLocalAddress() {
+    return transport.getLocalAddress();
+  }
+
+  @Override
+  public void close() throws Exception {
+    LOG.log(Level.FINE, "Shutting down");
+    this.nameServiceRegisteringStage.close();
+    this.nameServiceUnregisteringStage.close();
+    this.transport.close();
```

Comment:
`DefaultNetworkServiceImpl` uses `TransportFactory` to create a new `transport`, so it cannot be shared.